### PR TITLE
Flow resolution context to nested mappings

### LINKED ISSUE #2937: ResolutionContext lost in nested mapping
When using MapFrom(), any IMappingOperationOptions.Items that are passed in are not passed along to nested ResolutionContext.Mapper.Map() calls. I realize that I can then manually readd the Items to the nested Map() call, but I believe it should happen automatically.

(NOTE: The example below is extremely simplified. You may think, "these mappings should just be handled through _____ (e.g. conventions)", but just keep in mind the main point is that the IMappingOperationOptions.Items are lost.

### Source/destination types

```csharp
public class FromGarage
	{
		public List<FromCar> FromCars { get; set; }
	}

public class ToGarage
	{
		public List<ToCar> ToCars { get; set; }
	}

public class FromCar
	{
		public int Id { get; set; }
		public string Name { get; set; }
		public Door Door { get; set; }
	}

public class ToCar
	{
		public int Id { get; set; }
		public string Name { get; set; }
		public Door Door { get; set; }
	}
```

### Mapping configuration

```csharp
CreateMap<FromGarage, ToGarage>()
                .ForMember(dest => dest.ToCars, opts => opts.MapFrom((src, dest, destVal, ctx) => 
                {
                    var toCars= new List<ToCar>();

                    ToCar toCar;
                    foreach (var fromCar in src.FromCars)
                    {
// I KNOW I COULD PASS IN THE ITEM "DOOR" MANUALLY RIGHT HERE, BUT SHOULDN'T THIS HAPPEN AUTOMATICALLY?
                        toCar = ctx.Mapper.Map<ToCar>(fromCar);
                        if (toCar == null)
                            continue;

                        toCars.Add(toCar);
                    }

                    return toCars;
                }));

CreateMap<FromCar, ToCar>()
                .ConvertUsing((src, dest, ctx) =>
                {
                    ToCar toCar = null;

                    if (fromCar.Name != null)
                    {
                        toCar.Id= fromCar.Id;
                        toCar.Name = fromCar.Name;

                            // ERROR OCCURS HERE. ITEMS ARE NOT THERE.
                            var door = ((Door)ctx.Items["Door"]);

                        toCar.Door = door;
                    }

                    return toCar;
                });
```

### Expected behavior

I would expect that the nested "ctx.Mapper.Map" call would pass along the IMappingOperationOptions.Items that were passed to it via the ResolutionContext.

### Actual behavior

The ResolutionContext of the nested "ctx.Mapper.Map" call does not contain any of the Items that were originally passed in.

### Steps to reproduce

```csharp
var toGarage = _mapper.Map<ToGarage>(fromGarage, opts =>
            {
                opts.Items.Add("Door", _door);
            });
```

### PR DESCRIPTION
**Pull request by jbogard**
Fixes #2937

### CODE REVIEW

**Comment by lbargaoanu** [comment]
A minor version bump?

File: src/AutoMapper/Execution/RuntimeMapper.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            opts(_resolutionContext.Options);
+
+            return Map<TDestination>(source);
+        }
+
+        public TDestination Map<TSource, TDestination>(TSource source)
+            => _inner.Map(source, default(TDestination), _resolutionContext);
+
+        public TDestination Map<TSource, TDestination>(TSource source, Action<IMappingOperationOptions<TSource, TDestination>> opts)
+        {
```
[the remark this answers is not in the document]
Comment (jbogard) [reply src/AutoMapper/Execution/RuntimeMapper.cs]: I don't want to add any more optional parameters to the public API, if i can avoid it. Those have backwards compat problems.

The biggest problems were the overloads taking `Action<IMappingOperationOptions>`. Those really shouldn't be allowed inside an `IRuntimeMapper`.

One option would be to separate the interface/objects for `IMapper` and `IRuntimeMapper`. Only put the things you should be able to do from the outermost mapping API on `IMapper`, and only the things you can do during a runtime mapping on `IRuntimeMapper`.

File: src/AutoMapper/ResolutionContext.cs [hunk 2]
```diff
@@ -105,7 +106,7 @@ internal int GetTypeDepth(TypePair types)
         public ResolutionContext(IMappingOperationOptions options, IRuntimeMapper mapper)
         {
             Options = options;
-            Mapper = mapper;
+            Mapper = new RuntimeMapper(this, mapper);
```
[the remark this answers is not in the document]
Comment (jbogard) [reply src/AutoMapper/ResolutionContext.cs]: Ha! I've already introduced a separate object - I'm not sure it's a *terrible* idea. I'll push a commit to see what it looks like.

File: src/AutoMapper/Execution/RuntimeMapper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lbargaoanu) [reply src/AutoMapper/Execution/RuntimeMapper.cs]: That makes sense. It's weird right now.

File: src/AutoMapper/ResolutionContext.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbogard) [reply src/AutoMapper/ResolutionContext.cs]: OK just pushed this. I think it works better - and in the future, we may want to remove the implicit interface implementation and the `Mapper` property, and clean up the interface to only the things that make sense for doing mappings inside a `Map` context.

Also we may have to rename this object. `ResolutionContext` is....a weird name.

**Pull request merged**